### PR TITLE
Preconditions for checking equal size/blocksize/distribution + adjustments

### PR DESCRIPTION
**Pull request by albestro**
The main point of this PR is to add two preconditions for checking equality of the size/blocksize/distribution of two matrices:
- `DLAF_ASSERT_SIZE_EQ`
- `DLAF_ASSERT_BLOCKSIZE_EQ`
- `DLAF_ASSERT_DISTRIBUTED_EQ`

As a side-work, I tried to simplify the implementation of existing macros making full use of the API.

### CODE REVIEW

**Comment by albestro** [comment]
I report here output from some of them (ones not involved with grids)

```bash
# DLAF_ASSERT_SIZE_SQUARE(a);
[ERROR] /Users/ialberto/workspace/repos/DLA-Future/example2.cpp:38 : int main(int, char **)
(a.size().rows() == a.size().cols())
Matrix a is not square (10, 2)
```

```bash
# DLAF_ASSERT_BLOCKSIZE_SQUARE(a);
[ERROR] /Users/ialberto/workspace/repos/DLA-Future/example2.cpp:41 : int main(int, char **)
(a.blockSize().rows() == a.blockSize().cols())
Matrix a blocksize (2, 3) is not square
```

```bash
# DLAF_ASSERT_BLOCKSIZE_EQ(a, b);
[ERROR] /Users/ialberto/workspace/repos/DLA-Future/example2.cpp:44 : int main(int, char **)
(a.blockSize() == b.blockSize())
Blocksizes of matrix a and b are not the same ((2, 3) vs (2, 2))
```

```bash
# DLAF_ASSERT_MULTIPLIABLE_MATRICES(a, b, c, blas::Op::Trans, blas::Op::NoTrans);
[ERROR] /Users/ialberto/workspace/repos/DLA-Future/example2.cpp:47 : int main(int, char **)
rows(mat_a.size(), opA) == mat_c.size().rows() && cols(mat_a.size(), opA) == rows(mat_b.size(), opB) && cols(mat_b.size(), opB) == mat_c.size().cols()
Size mismatch: a (2, 10) x b (2, 4) --> c (10, 4) cannot be performed
```

```bash
# DLAF_ASSERT_MULTIPLIABLE_MATRICES(a, b, c, blas::Op::NoTrans, blas::Op::NoTrans);
[ERROR] /Users/ialberto/workspace/repos/DLA-Future/example2.cpp:50 : int main(int, char **)
rows(mat_a.blockSize(), opA) == mat_c.blockSize().rows() && cols(mat_a.blockSize(), opA) == rows(mat_b.blockSize(), opB) && cols(mat_b.blockSize(), opB) == mat_c.blockSize().cols()
Blocksize mismatch: a (2, 3) x b (2, 2) --> c (2, 2) cannot be performed
```

File: include/dlaf/util_matrix.h [hunk 1]
```diff
@@ -39,40 +39,61 @@ namespace internal {
...
 #define DLAF_ASSERT_SIZE_SQUARE(matrix)                                                               \
-  DLAF_ASSERT((matrix.size().rows() == matrix.size().cols()), "Matrix ", #matrix, " is not square (", \
-              matrix.size().rows(), "x", matrix.size().cols(), ").")
+  DLAF_ASSERT((matrix.size().rows() == matrix.size().cols()), "Matrix ", #matrix, " ", matrix.size(), \
+              " is not square")
+
+/// @brief Assert that @p matrixA and @p matrixB have the same size.
+///
+/// When the assertion is enabled, terminates the program with an error message if the two
+/// matrices does not have the same size. This assertion is enabled when **DLAF_ASSERT_ENABLE** is ON.
```
Comment (Sely85) [inline include/dlaf/util_matrix.h:48]:
```suggestion
/// matrices do not have the same size. This assertion is enabled when **DLAF_ASSERT_ENABLE** is ON.
```

File: include/dlaf/util_matrix.h [hunk 2]
```diff
@@ -39,40 +39,61 @@ namespace internal {
...
-              #matrix, " is not square (", matrix.blockSize().rows(), "x", matrix.blockSize().cols(), \
-              ").")
+#define DLAF_ASSERT_BLOCKSIZE_SQUARE(matrix)                                                \
+  DLAF_ASSERT((matrix.blockSize().rows() == matrix.blockSize().cols()), "Matrix ", #matrix, \
+              " blocksize ", matrix.blockSize(), " is not square")
+
+/// @brief Assert that @p matrixA and @p matrixB tiles have the same size.
+///
+/// When the assertion is enabled, terminates the program with an error message if the blocksize of the two
+/// matrices does not have the same size. This assertion is enabled when **DLAF_ASSERT_ENABLE** is ON.
```
Comment (Sely85) [inline include/dlaf/util_matrix.h:64]:
```suggestion
/// matrices do not have the same size. This assertion is enabled when **DLAF_ASSERT_ENABLE** is ON.
```

**Comment by teonnik** [comment]
This looks good to me but if I may suggest (in a new PR) to extend the ASSERT mechanism to handle custom handlers which control printing the message so that instead of this:

```
#define DLAF_ASSERT_DISTRIBUTED_ON_GRID(grid, matrix)                                          \
  DLAF_ASSERT(((matrix.commGridSize() == grid.size()) && (matrix.rankIndex() == grid.rank())), \
              "The matrix ", #matrix, " (rank: ", matrix.rankIndex(),                          \
              ", grid size: ", matrix.commGridSize(),                                          \
              ") is not distributed according to the communicator grid ", #grid,               \
              " (rank: ", grid.rank(), ", grid size: ", grid.size(), ").")
```

we write this  (pseudo-code-ish):

```
bool assert_distributed_on_grid(Grid grid, Matrix matrix) {
    return  matrix.commGridSize() == grid.size() && matrix.rankIndex() == grid.rank();
}

str::string assert_distributed_on_grid_handler(const char* condition, Grid grid, Matrix matrix, const char* grid_name, const char* matrix_name) {
   std::stringstream ss;
   ss  << condition <<
        <<   "The matrix " << matrix_name << " (rank: ", matrix.rankIndex(),"
        <<  ", grid size: " << matrix.commGridSize()  << ") is not distributed according to the communicator grid "
       <<  #grid <<  " (rank: " << grid.rank() <<  ", grid size: " << grid.size() << ").");
  return ss.str();
}

...

DLAF_ASSERT(assert_distributed_on_grid(matrix, grid),  assert_distributed_on_grid_handler, grid, matrix);
```

This way, there is no need for new macros and most functions can be regular C++. 

To achieve that our ASSERT_ will need to be able to accept a custom handler. Currently we have this (in DLAF_CHECK_WITH_ORIGIN) :

```
    std::cerr << "[ERROR] " << origin << std::endl                    \
                << dlaf::common::concat(#condition, '\n', __VA_ARGS__); \
```

which should probably change to: 

```
std::cerr << "[ERROR] " << origin << std::endl                    \
                << handler(#condition,  __VA_ARGS__,  #__VA_ARGS__); \

```

**Comment by rasolca** [comment]
bors try

**Pull request merged**